### PR TITLE
Print ersilia catalog in a table natively 

### LINKED ISSUE #1268: 🐈 Task: Print ersilia catalog in a table natively
### Summary

Related to #12 #238 

Previous volunteers to ersilia did some fantastic work in making Ersilia catalog display in a table format on the terminal, however the external dependency used to achieve that functionality caused ersilia environment to be bloated. We do not want this to be the case and therefore want to implement this functionality natively in Python.

### Objective(s)

_No response_

### Documentation

_No response_

### PR DESCRIPTION
**Pull request by jubril-adeyi**
This PR adds a new method to the CatalogTable class to present the model catalog in a native table format without external dependencies.

**Changes to be made**

- Implemented the as_table method in the CatalogTable class to generate a formatted table representation of the data, complete with borders and proper alignment.

- Updated the existing __str__ method to return the table format when calling str() on CatalogTable objects.
- Updated the catalog_md function in /cli/command/catalog to call the as_table method from the as_json method
**Status**
I have completed the implementation of the as_table method, ensuring that it handles data formatting and alignment correctly. 

**To do**
- Review the implementation for any potential optimizations or edge cases.

Closes #1268

### CODE REVIEW

**Comment by jubril-adeyi** [comment]
@GemmaTuron @DhanshreeA

I have applied the changes locally and tested the output of the catalog --more command. The table now displays properly formatted data with borders and row separators. I have attached the output from my tests for your review.

image

image

File: ersilia/hub/content/catalog.py [hunk 1]
```diff
@@ -28,19 +28,29 @@ class CatalogTable(object):
...
+        table = "┌" + "┬".join("─" * (width + 2) for width in column_widths) + "┐\n" 
+        table += "│ " + row_format.format(*self.columns) + " │\n"  
+        table += "├" + "┼".join("─" * (width + 2) for width in column_widths) + "┤\n"
+
+        for index, row in enumerate(self.data):
+            row = [str(item) if item is not None else "" for item in row]
+            table += "│ " + row_format.format(*row) + " │\n"
+
+            if index < len(self.data) - 1:
+                table += "├" + "┼".join("─" * (width + 2) for width in column_widths) + "┤\n"
```
Comment (DhanshreeA) [inline ersilia/hub/content/catalog.py:48]: Can we make these border characters as importable constants in the defaults.py file? Ideally we should not have strings as much as possible in the code.

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

**Comment by DhanshreeA** [comment]
@jubril-adeyi fantastic work! Can you please try this out with a couple of more models - you can browse them in ersilia.io page. I want to see how it behaves with models that have longer titles.

File: ersilia/hub/content/catalog.py (hunk 1 above)
Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py:48]: Yes, and I agree. I'll implement this change and update the PR shortly

**Comment by jubril-adeyi** [comment]
Thank you @DhanshreeA ,  The current logic sets the table column widths based on the longests value in each column, so it handles models with longer titles by making the column wide enough. I've attached a .txt file containing the output of the 'ersilia catalog --hub --more' command, which includes some models with longer titles
[hubcatalog.txt](https://github.com/user-attachments/files/17267699/hubcatalog.txt)

File: ersilia/hub/content/catalog.py (hunk 1 above)
Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py]: @DhanshreeA the change has been implemented. border characters and are now set as constants in the defaults.py file

**Comment by Malikbadmus** [comment]
Amazing work @jubril-adeyi , just requested some minor changes.

**Comment by Malikbadmus** [comment]
@DhanshreeA ,I've tested the code, and everything is working as expected. It looks good to be merged.

**Review by Malikbadmus — CHANGES_REQUESTED** [review]

File: ersilia/hub/content/catalog.py [hunk 2]
```diff
@@ -28,19 +33,29 @@ class CatalogTable(object):
...
-                d[c] = r[i]
-            R += [d]
-        return R
+        table = TABLE_TOP_LEFT + TABLE_TOP_MIDDLE.join(TABLE_HORIZONTAL * (width + 2) for width in column_widths) + TABLE_TOP_RIGHT + "\n"
+        table += TABLE_VERTICAL + " " + row_format.format(*self.columns) + " " + TABLE_VERTICAL + "\n"
+        table += TABLE_MIDDLE_LEFT + TABLE_MIDDLE_MIDDLE.join(TABLE_HORIZONTAL * (width + 2) for width in column_widths) + TABLE_MIDDLE_RIGHT + "\n"
+
+        for index, row in enumerate(self.data):
+            row = [str(item) if item is not None else "" for item in row]
+            table += TABLE_VERTICAL + " " + row_format.format(*row) + " " + TABLE_VERTICAL + "\n"
```
[the remark this answers is not in the document]
Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py]: @Malikbadmus This change has been implemented and added to this PR

File: ersilia/hub/content/catalog.py [hunk 3]
```diff
@@ -28,19 +33,29 @@ class CatalogTable(object):
...
+
+            if index < len(self.data) - 1:
+                table += TABLE_MIDDLE_LEFT + TABLE_MIDDLE_MIDDLE.join(TABLE_HORIZONTAL * (width + 2) for width in column_widths) + TABLE_MIDDLE_RIGHT + "\n"
+
+        table += TABLE_BOTTOM_LEFT + TABLE_BOTTOM_MIDDLE.join(TABLE_HORIZONTAL * (width + 2) for width in column_widths) + TABLE_BOTTOM_RIGHT
+        
+        return table
```
[the remark this answers is not in the document]
Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py:61]: @Malikbadmus  I’ve implemented a helper function,, to handle the repeated logic for constructing the  separators, please let me know if you have further feedback.

**Review by Malikbadmus — APPROVED** [review]

File: ersilia/cli/commands/catalog.py [hunk 4]
```diff
@@ -53,7 +53,7 @@ def catalog(local=False, file_name=None, browser=False, more=False):
             )
             return
         if file_name is None:
-            catalog = catalog_table.as_json()
+            catalog = catalog_table.as_table()
```
Comment (DhanshreeA) [inline ersilia/cli/commands/catalog.py:56]: I don't think we should change this from `as_json` to `as_table`. We should instead have a flag called `--as-table` and check if it is set, and only then return the result as a table.

File: ersilia/default.py [hunk 5]
```diff
@@ -104,6 +104,21 @@
...
+TABLE_HORIZONTAL = "─"
+TABLE_VERTICAL = "│"
+TABLE_MIDDLE_LEFT = "├"
+TABLE_MIDDLE_MIDDLE = "┼"
+TABLE_MIDDLE_RIGHT = "┤"
+TABLE_BOTTOM_LEFT = "└"
+TABLE_BOTTOM_MIDDLE = "┴"
+TABLE_BOTTOM_RIGHT = "┘"
+TABLE_CELL_PADDING = " "
+COLUMN_SEPARATOR = " | "
```
Comment (DhanshreeA) [inline ersilia/default.py:120]: Let's create a string enum class that has these as attributes. This way when you import it in `ersilia/hub/content/catalog.py`, there are not a lot of imports to deal with.

File: ersilia/hub/content/catalog.py [hunk 6]
```diff
@@ -28,19 +33,33 @@ class CatalogTable(object):
...
+            table += TABLE_VERTICAL + TABLE_CELL_PADDING + row_format.format(*row) + TABLE_CELL_PADDING + TABLE_VERTICAL + "\n"
+
+            if index < len(self.data) - 1:
+                table += self.generate_separator_line(TABLE_MIDDLE_LEFT, TABLE_MIDDLE_MIDDLE, TABLE_MIDDLE_RIGHT, TABLE_HORIZONTAL, column_widths) + "\n"
+
+        table += self.generate_separator_line(TABLE_BOTTOM_LEFT, TABLE_BOTTOM_MIDDLE, TABLE_BOTTOM_RIGHT, TABLE_HORIZONTAL, column_widths)
+    
+        return table 
 
-    def as_json(self):
```
Comment (DhanshreeA) [inline ersilia/hub/content/catalog.py:41]: Let's please keep this method.

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

**Comment by DhanshreeA** [comment]
> Thank you @DhanshreeA , The current logic sets the table column widths based on the longests value in each column, so it handles models with longer titles by making the column wide enough. I've attached a .txt file containing the output of the 'ersilia catalog --hub --more' command, which includes some models with longer titles [hubcatalog.txt](https://github.com/user-attachments/files/17267699/hubcatalog.txt)

The output looks fantastic, thanks @jubril-adeyi! Minor clean up requested.

File: ersilia/hub/content/catalog.py (hunk 6 above)
Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py:41]: Okay, I've re-added the as_json method as requested in my latest commit

Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py:41]: @DhanshreeA I have re-added the as_json method in the latest commit.

File: ersilia/cli/commands/catalog.py (hunk 4 above)
Comment (jubril-adeyi) [reply ersilia/cli/commands/catalog.py:56]: @DhanshreeA Thanks for the feedback,  and I agree , the json output as the default, and the table format output when specified makes a lot more sense.
I will implement the flag so the output can be conditionally formatted as a table when specified.

Comment (jubril-adeyi) [reply ersilia/cli/commands/catalog.py:63]: @DhanshreeA I have implemented the --as-table flag in the latest commits. Now the default behaviour remains the same, returning the json output, unless the --as-table flag is added to the command.

File: ersilia/default.py (hunk 5 above)
Comment (jubril-adeyi) [reply ersilia/default.py]: @DhanshreeA I’ve implemented this in the latest commit,created a TableConstants enum class that handles all the table formatting attributes.

File: ersilia/hub/content/catalog.py [hunk 7]
```diff
@@ -37,10 +38,40 @@ def as_list_of_dicts(self):
...
+    
     def as_json(self):
         R = self.as_list_of_dicts()
         return json.dumps(R, indent=4)
+    
+    def generate_separator_line(self, left, middle, right, horizontal, widths):
+        return left + middle.join(horizontal * (width + 2) for width in widths) + right
+
+    def as_table(self):
+        column_widths = [
```
Comment (DhanshreeA) [inline ersilia/hub/content/catalog.py:50]: @jubril-adeyi apologies for dragging this out - just few minor things before we close this PR:
Could you add a docstring for the `as_table` method to explain your logic?
Also please format the code with black - this is extending far beyond the ideal 80 column width for python programs.

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

**Comment by DhanshreeA** [comment]
@jubril-adeyi could you also please resolve the merge conflicts in this PR?

File: ersilia/hub/content/catalog.py (hunk 7 above)
Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py:86]: @DhanshreeA docstring has been added for the as_table method and the seperator_line helper method, explaining the logic. 
The code has also been formated using black, in the latest commit

**Comment by jubril-adeyi** [comment]
@DhanshreeA The merge conflicts in the PR has been resolved

File: ersilia/cli/commands/catalog.py [hunk 8]
```diff
@@ -80,6 +86,21 @@ def catalog(
             mc.airtable()
             return
         catalog_table = mc.local() if local else mc.hub()
+        if local is True and browser is True:
```
Comment (DhanshreeA) [inline ersilia/cli/commands/catalog.py:89]: This code from L89 to L102 is repeating, ie it's the same as code from L75 till L88

File: ersilia/cli/commands/catalog.py [hunk 9]
```diff
@@ -88,10 +109,13 @@ def catalog(
                 )
             )
             return
-        if file_name is None:
-            catalog = catalog_table.as_json()
+        if as_table:
```
Comment (DhanshreeA) [inline ersilia/cli/commands/catalog.py:112]: I think this logic is between whether we want to show the catalog on the screen vs save it to a file. So, we should indeed have the following flow:

```Python
if file_name is None:
     catalog = catalog_table.as_table() if as_table else catalog_table.as_json()
else:
     catalog_table.write(file_name)
     catalog = None
```

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

File: ersilia/hub/content/catalog.py [hunk 10]
```diff
@@ -122,7 +239,7 @@ def airtable(self):
             webbrowser.open("https://airtable.com/shrUcrUnd7jB9ChZV")
 
     def _get_all_github_public_repos(self):
-        url = "https://api.github.com/users/{0}/repos".format(GITHUB_ORG)
+        url = "https://api.github.com/users/y{0}/repos".format(GITHUB_ORG)
```
Comment (DhanshreeA) [inline ersilia/hub/content/catalog.py:242]: There's an extraneous `y` here. This is causing tests to fail because it's forming an incorrect URL.

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

File: ersilia/cli/commands/catalog.py (hunk 8 above)
Comment (jubril-adeyi) [reply ersilia/cli/commands/catalog.py:89]: I must have missed that while resolving the merge conflict, I'll fix this

File: ersilia/cli/commands/catalog.py (hunk 9 above)
Comment (jubril-adeyi) [reply ersilia/cli/commands/catalog.py:112]: @DhanshreeA Thanks, I've implemented this is the latest commit

File: ersilia/hub/content/catalog.py (hunk 10 above)
Comment (jubril-adeyi) [reply ersilia/hub/content/catalog.py]: @DhanshreeA apologies for this error, his has also been resolved in the latest commit

**Review by DhanshreeA — APPROVED** [review]

**Pull request merged**